src/rapids_pre_commit_hooks/alpha_spec.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import reduce
+
+import yaml
+from packaging.requirements import Requirement
+from packaging.specifiers import SpecifierSet
+
+from .lint import LintMain
+
+RAPIDS_VERSIONED_PACKAGES = {
+    "rmm",
+    "pylibcugraphops",
+    "pylibcugraph",
+    "nx-cugraph",
+    "dask-cudf",
+    "cuspatial",
+    "cuproj",
+    "cuml",
+    "cugraph",
+    "cudf",
+    "ptxcompiler",
+    "cubinlinker",
+    "cugraph-dgl",
+    "cugraph-pyg",
+    "cugraph-equivariant",
+    "raft-dask",
+    "pylibwholegraph",
+    "pylibraft",
+    "cuxfilter",
+    "cucim",
+    "ucx-py",
+    "ucxx",
+    "pynvjitlink",
+    "distributed-ucxx",
+}
+
+ALPHA_SPECIFIER = ">=0.0.0a0"
+
+ALPHA_SPEC_OUTPUT_TYPES = {
+    "pyproject",
+    "requirements",
+}
+
+
+def check_package_spec(linter, args, node):
+    if node.tag == "tag:yaml.org,2002:str":
+        req = Requirement(node.value)
+        if req.name in RAPIDS_VERSIONED_PACKAGES:
+            has_alpha_spec = any(
+                filter(lambda s: str(s) == ALPHA_SPECIFIER, req.specifier)
+            )
+            if args.mode == "development" and not has_alpha_spec:
+                req.specifier &= ALPHA_SPECIFIER
+                linter.add_warning(
+                    (node.start_mark.index, node.end_mark.index),
+                    f"add alpha spec for RAPIDS package {req.name}",
+                ).add_replacement(
+                    (node.start_mark.index, node.end_mark.index), str(req)
+                )
+            elif args.mode == "release" and has_alpha_spec:
+                req.specifier = reduce(
+                    lambda ss, s: ss & str(s),
```

Comment:
Why do we need to merge them as strings here? I thought we would want to use the SpecifierSet union operator? Or is this to preserve some sort of ordering guarantee that we wouldn't get with the SpecifierSet?

Reply:
LOL yes that is silly. OK that's fine.